lib/oauth2-client.js

```diff
@@ -0,0 +1,185 @@
+const {util: {BedrockError}} = bedrock;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await database.openCollections(['access-tokens']);
+
+  await database.createIndexes([{
+    // prevent duplicate tokens from being stored
+    collection: 'access-tokens',
+    fields: {'token.token': 1},
+    options: {unique: true, background: false}
+  }, {
+    // automatically expire tokens with an `expires` date field
+    collection: 'access-tokens',
+    fields: {'token.expires': 1},
+    options: {
+      partialFilterExpression: {'token.expires': {$exists: true}},
+      unique: false,
+      background: false,
+      expireAfterSeconds: 0
+    }
+  }]);
+});
+
+export class OAuth2Client {
+/**
+ * Constructs a OAuth2Client instance that can be used to return an accessToken.
+ *  It first checks the database to see if an accessToken already exists for the
+ *  client id; if not it creates an accessToken then stores it in the database.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.config - The "oauth2-client-grant" properties.
+ * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+ *
+ * @returns {OAuth2Client} The `OAuth2Client` instance.
+ */
+  constructor({config, httpsAgent} = {}) {
+    this.client_id = config.client_id;
+    this.client_secret = config.client_secret;
+    this.token_endpoint = config.token_endpoint;
+    this.grant_type = config.grant_type;
+    this.scope = config.scope;
+    this.maxRetries = config.maxRetries;
+    this.httpsAgent = httpsAgent;
+  }
+  /**
+   * Gets a new accessToken by first checking the database to see if an
+   *  accessToken already exists for the client id; if not it creates an
+   *  accessToken then stores it in the database.
+   *
+   * @returns {string} The accessToken.
+   */
+  async getToken() {
+    const {
+      client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+      maxRetries
+    } = this;
+    let token;
+    token = await checkDatabase({client_id});
+    if(!token || !token.access_token) {
+      token = await getNewAccessToken({
+        client_id, client_secret, token_endpoint, grant_type, scope, httpsAgent,
+        maxRetries
+      });
+      await insertToken({token});
+    }
+    return token.access_token;
+  }
+}
+
+/**
+ * Gets a new accessToken from the provided URL.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.client_id - The ID of the client.
+ * @param {string} options.client_secret - The client secret.
+ * @param {string} options.token_endpoint - The URL to call.
+ * @param {string} options.grant_type - The grant type.
+ * @param {string} options.scope - The scope.
+ * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+ * @param {number} options.maxRetries - The maximum number of times to retry
+ *  the request.
+ *
+ * @returns {string} The accessToken.
+ */
+async function getNewAccessToken({
```

Comment:
Move all non-exported helper functions below all exported functions so that someone reading the file will hit the public exports first.